plugins/module_utils/ansible_freeipa_module.py

```diff
@@ -309,3 +349,159 @@ def is_ipv6_addr(ipaddr):
     except socket.error:
         return False
     return True
+
+
+class AnsibleFreeIPAParams(dict):
+    def __init__(self, ansible_module):
+        self.update(ansible_module.params)
+        self.ansible_module = ansible_module
+
+    @property
+    def names(self):
+        return self.name
+
+    def __getattr__(self, name):
+        param = self.get(name)
+        if param is not None:
+            return _afm_convert(param)
+
+
+class FreeIPABaseModule(AnsibleModule):
+    ipa_param_mapping = {}
+
+    def __init__(self, *args, **kwargs):
+        super(FreeIPABaseModule, self).__init__(*args, **kwargs)
+
+        self.ccache_dir = None
+        self.ccache_name = None
+        self.changed = False
+        self.exit_args = {}
+        self.ipa_connected = False
+        self._debug = True
+
+        self.ipa_params = AnsibleFreeIPAParams(self)
+        self.ipa_commands = []
+
+    def get_ipa_command_args(self):
+        args = {}
+        for ipa_param_name, param_name in self.ipa_param_mapping.items():
+            if param_name in self.ipa_params:
+                value = self.ipa_params.get(param_name)
+                if isinstance(value, bool):
+                    value = "TRUE" if value else "FALSE"
+
+            elif hasattr(self, param_name):
+                method = getattr(self, param_name)
+                if callable(method):
+                    value = method()
+            else:
+                value = None
+
+            if value is not None:
+                args[ipa_param_name] = value
+
+        return args
+
+    def check_ipa_params(self):
+        pass
+
+    def define_ipa_commands(self):
+        pass
+
+    def api_command(self, command, name=None, args=None):
+        if args is None:
+            args = {}
+
+        if name is None:
+            return api_command_no_name(self, command, args)
+
+        return api_command(self, command, name, args)
```

Comment:
Wouldn't it be better to reimplement the api_command* functions here using *args and **kwargs?
I mean, there are a few API commands that can take more the one positional parameter, and with variable arguments, this calls would also be contemplated.
As we spoke before, I wanted to test it, but could not do it so far.

Reply:
I've added this commit to test that:

https://github.com/freeipa/ansible-freeipa/pull/224/commits/632e1e0c1842c1ef4a1fa5e4e0cb1cf39e831c6e

I'm not sure I like it. It works and I'm ok with the kwargs version I'm just not sure about replacing name by *args. It think it was a bit harder to understand and also seems to be masking the actual parameter.

You mentioned that a few commands can take more than one positional parameter, right? Could you point me to those so I can take a look and see if I can think of an better proposal?

Reply:
After discussing on video call we decided to drop this idea for now.